skeleton/.gitlab-ci.yml

```diff
@@ -1,63 +1,275 @@
+  BUNDLER_VERSION: '1.16.1'
+  MODULE_NAME: "$(ruby -r json -e 'JSON.load(File.read('metadata.json'))['name']')"
+  MODULE_VERSION: "$(ruby -r json -e 'JSON.load(File.read('metadata.json'))['version']')"
+
+before_script:
+  - 'ruby -v'
+  - 'gem list -ie --silent -v "$BUNDLER_VERSION" bundler && bundle -v || gem install bundler --no-document -v "$BUNDLER_VERSION"'
+  - 'bundle install -j $(nproc) --no-binstubs --path vendor'
+
 cache:
+  key: '${CI_COMMIT_REF_SLUG}'
   paths:
-    - vendor
-    - .bundle
-puppet-syntax:
-  stage: test
-  tags:
-    - docker
-  image: ruby:2.1
+    - 'vendor/ruby'
+
+#=======================================================================
+# Anchors
+
+.lint_base: &lint_base
+  stage: 'lint'
+  tags: ['docker']
   script:
-    - bundle install
-    - bundle exec rake syntax
-puppet-lint:
-  stage: test
-  tags:
-    - docker
-  image: ruby:2.1
+    - 'bundle exec rake syntax'
+    - 'bundle exec rake lint'
+  cache:
+    policy: 'pull'
+  dependencies: []
+  artifacts:
+    when: 'always'
+    paths:
+      - 'Gemfile.lock'
+
+.unit_base: &unit_base
+  stage: 'unit'
+  tags: ['docker']
   script:
-    - bundle install
-    - bundle exec rake lint
-puppet-metadata:
-  stage: test
-  tags:
-    - docker
-  image: ruby:2.1
+    - 'bundle exec rake spec'
+  cache:
+    policy: 'pull'
+  dependencies: []
+  artifacts:
+    when: 'always'
+    paths:
+      - 'Gemfile.lock'
+
+.acceptance_base: &acceptance_base
+  stage: 'acceptance'
+  tags: ['beaker']
+  cache:
+    policy: 'pull'
+  dependencies: []
+  artifacts:
+    when: 'always'
+    paths:
+      - 'Gemfile.lock'
+
+# ----------------------------------------------------------------------
+# Version Matrix
+#
+# It would be too expensive, both in time and compute resources, to test
+# against every last version combination, so we restrict it to this subset.
+# Version sets are selected based on current support policies for major platform
+# software, such as Puppet and Ruby.  Namely, we use the version combinations
+# bundled in Puppet Enterprise.
+#
+# For more information see:
+#  * https://puppet.com/docs/pe/latest/overview/component_versions_in_recent_pe_releases.html
+#  * https://puppet.com/misc/puppet-enterprise-lifecycle
+#  * https://puppet.com/docs/pe/latest/overview/getting_support_for_pe.html#standard-releases-and-long-term-support-releases
+#
+# ----------------------------------------------
+# | Release | Puppet | Ruby | End-of-Life Date |
+# ----------|--------|------|------------------|
+# PE 2016.4   4.7      2.1    2018-10
+# PE 2016.5   4.8      2.1    2017-05
+# SIMP 6.0    4.8      2.1    TBD
+# PE 2017.1   4.9      2.1    2017-10
+# PE 2017.2   4.10     2.1    2018-02
+# PE 2017.3   5.3      2.4    2018-08
+# PE 2018.1   5.5      2.4    2020-05
+#
+
+.pe_2016_4: &pe_2016_4
+  variables:
+    PUPPET_VERSION: '~> 4.10.0'
+    BEAKER_PUPPET_COLLECTION: 'pc1'
+    BEAKER_PUPPET_AGENT_VERSION: '~> 1.10.0'
+
+.pe_2016_5: &pe_2016_5
```

Comment:
According to the handy chart in the comments, we can drop tests for:

* PE 2016.5
* PE 2017.1 
* PE 2017.2

This checks out with the latest info at https://puppet.com/misc/puppet-enterprise-lifecycle, which lists the current active PE versions as:

Active Puppet Enterprise versions | Start mainstream support | Start extended support | End of life
-- | -- | -- | --
PE 2016.4 (LTS) | 20 October 2016 | April 2018 | October 2018
PE 2017.2 | 11 May 2017 | 11 November 2017 | February 2018
PE 2017.3 | October 2017 | May 2018 | August 2018
PE 2018.1 (LTS)